Nay chi than shwe/feature/reimplement constraint form

naychithanshwe opened the pull request:
## Describe your changes
Created View and View Model for scheduler constraint form 
## Issue ticket number and link

naychithanshwe committed:
new file AddShiftConstraintsForm created in Shfit_Constraints directory

naychithanshwe committed:
new file AddShiftConstraintsForm created in Shfit_Constraints directory

naychithanshwe committed:
set schedulerForm route in main.dart

naychithanshwe committed:
set schedulerForm route in main.dart

Commit:
Fix form reference

Commit:
Fix form reference

naychithanshwe committed:
scheduler form route added to main.dart

naychithanshwe committed:
scheduler form route added to main.dart

naychithanshwe committed:
ConstraintForm.dart and ConstraintFormViewModel.dart classes created

naychithanshwe committed:
ConstraintForm.dart and ConstraintFormViewModel.dart classes created

naychithanshwe committed:
dropdown menu added to ConstraintForm.dart

naychithanshwe committed:
dropdown menu added to ConstraintForm.dart

naychithanshwe committed:
internationalization applied for label texts

naychithanshwe committed:
internationalization applied for label texts

naychithanshwe committed:
dropdown menu entry fixed

naychithanshwe committed:
dropdown menu entry fixed

naychithanshwe committed:
textbox alignment fixed

naychithanshwe committed:
textbox alignment fixed

naychithanshwe committed:
Datebox fixed

naychithanshwe committed:
Datebox fixed

naychithanshwe committed:
Calendar icon added to date box

naychithanshwe committed:
Calendar icon added to date box

naychithanshwe committed:
Name box done

naychithanshwe committed:
Name box done

naychithanshwe committed:
timebox text field style changed

naychithanshwe committed:
timebox text field style changed

naychithanshwe committed:
all textbox field styles fixed

naychithanshwe committed:
all textbox field styles fixed

naychithanshwe committed:
text colors changed and added sizedbox

naychithanshwe committed:
text colors changed and added sizedbox

naychithanshwe committed:
fix widgets

naychithanshwe committed:
fix widgets

naychithanshwe committed:
add schedule button centered

naychithanshwe committed:
add schedule button centered

naychithanshwe committed:
button fixed

naychithanshwe committed:
button fixed

naychithanshwe committed:
comments added

naychithanshwe committed:
comments added

naychithanshwe committed:
Reimplement Constraint Form Branch

naychithanshwe committed:
constraint form route moved

naychithanshwe committed:
Merge remote-tracking branch 'origin/main' into NayChiThanShwe/Feature/Reimplement-constraint-form

naychithanshwe committed:
delete unnecessary AddShiftConstraintsForm

naychithanshwe committed:
delete unnecessary AddShiftConstraintsForm

naychithanshwe committed:
changed name for constraint_form.dart and constraint_form_view_model.dart

naychithanshwe committed:
changed name for constraint_form.dart and constraint_form_view_model.dart

naychithanshwe committed:
constraint_form route added to main

naychithanshwe committed:
constraint_form route added to main

naychithanshwe committed:
border variable created

naychithanshwe committed:
border variable created

naychithanshwe committed:
view model created

naychithanshwe committed:
view model created

naychithanshwe committed:
submitForm added in view model

naychithanshwe committed:
submitForm added in view model

naychithanshwe committed:
dispose function added to view model

naychithanshwe committed:
dispose function added to view model

naychithanshwe committed:
submitForm fixed

naychithanshwe committed:
submitForm fixed

naychithanshwe committed:
parent and child classes created

naychithanshwe committed:
parent and child classes created

naychithanshwe committed:
const added

naychithanshwe committed:
const added

naychithanshwe committed:
split class

naychithanshwe committed:
split class

naychithanshwe committed:
renamed class

naychithanshwe committed:
renamed class

naychithanshwe committed:
Merge remote-tracking branch 'origin/main' into NayChiThanShwe/Feature/Reimplement-constraint-form

emilymclean commented on fireapp/lib/l10n/app_en.arb:
```diff
@@ -123,4 +123,10 @@
   "serverUrlBlurb": "You need to enter the URL of a development environment",
   "serverUrlHint": "Server URL",
   "serverUrlRemove": "This page can be skipped in official builds by running \"dart run environment_config:generate --serviceUrl=<your api url>\" before building."
```
You need to add a comma after this. Currently your branch won't build

emilymclean commented on fireapp/lib/presentation/login/login_view_model.dart:
```diff
@@ -26,7 +26,7 @@ class LoginViewModel
   bool obscureText = true;
 
-  final BehaviorSubject<LoginNavigation> _navigate = BehaviorSubject();
+  final BehaviorSubject<LoginNavigation>  _navigate = BehaviorSubject();
```
Can you undo this change? This seems like a mistake

emilymclean commented on fireapp/lib/presentation/register/register_page.dart:
```diff
@@ -59,7 +59,7 @@ class _RegisterPageState
             SizedBox(height: 1.rdp(),),
             FillWidth(
               child: Text(
-                AppLocalizations.of(context)?.registerTitle ?? "",
```
Undo this please

emilymclean commented on fireapp/lib/presentation/constraint_form/constraint_form_view.dart:
```diff
@@ -0,0 +1,186 @@
+
+enum schedulerInputType { text, date, time }
+
+class ConstraintFormRoute extends StatelessWidget {
```
Please rename this to "ScheduleConstraintPage"

emilymclean commented on fireapp/lib/presentation/constraint_form/constraint_form_view.dart:
```diff
@@ -0,0 +1,186 @@
+  }
+}
+
+class ConstraintForm extends StatefulWidget {
```
Like above, rename to "ScheduleConstraintForm"

emilymclean commented on fireapp/lib/presentation/constraint_form/constraint_form_view_model.dart:
```diff
@@ -0,0 +1,41 @@
+  Stream<RequestState<void>> get state => _state.stream;
+
+  // Dummy values for the assets list
+  final Stream<List<int>> assetsStream = Stream<List<int>>.value([1, 2, 3, 4]);
```
This should be a list of Asset objects (which I forgot to ask you to write). I'll make a PR into your branch with a model, then please update this.

emilymclean replied:
PR #48 contains this model

emilymclean commented on fireapp/lib/presentation/constraint_form/constraint_form_view_model.dart:
```diff
@@ -0,0 +1,41 @@
+
+  // Function to be called when the form is submitted
+  void submitForm() {
+    if (formKey.currentState?.validate() == true) {
```
I prefer early return to this style. This would mean you change it to:
``` dart
void submitForm() {
  if (formKey.currentState?.validate() != true) return;
  // Submission logic here
}
```

emilymclean commented on fireapp/lib/presentation/constraint_form/base_input_field.dart:
```diff
@@ -0,0 +1,103 @@
+import 'package:flutter/cupertino.dart';
```
I can see what you're doing with this class, but I think you're approaching it wrong. In my opinion "BaseInputField" should be a simple wrapper on TextField that applies styling and exposes many fields for configuring that styling. You would then have classes extending it that apply specific, standard, styles for each type of data (time, date, text).

I would have the structure of:
```
        [ BaseInputField ]
        |                |
        v                v
[TemporalInputField] [TextInputField]
|                 |
v                 v
[TimeInputField]  [DateInputField]
```

TemporalInputField would not expose a TextEditingController, but a value field and onValueChanged callback. TimeInputField and DateInputField would just call their respective pickers. I suggest this because I think it is bad practice to pass around string representations of dates.

emilymclean commented on fireapp/lib/presentation/constraint_form/constraint_form_view.dart:
```diff
@@ -0,0 +1,186 @@
+        );
+}
+
+class _SchedulerDateInput extends BaseInputField {
```
These _Scheduler...Input classes are unnecessary.

emilymclean commented on fireapp/lib/presentation/constraint_form/constraint_form_view.dart:
```diff
@@ -0,0 +1,186 @@
+        );
+}
+
+var commonInputBorder = const OutlineInputBorder(
```
Avoid these kinds of constants. Just put this in the BaseInputField as a private member for now.

emilymclean reviewed (changes requested):
This is looking good, but I think you should approach the input fields differently.

Commit:
Add asset type model & data layer

This commit adds functionality for retrieving asset types from the server. It includes changes to the REST client, reference data client, domain models, and repository. The new `AssetType` model is introduced along with its associated JSON serialization and deserialization logic. The `getAssetTypes` method is added to both the REST client and reference data client to fetch asset types from the server. Finally, the `getAssetType` method is implemented in the reference data repository to retrieve a list of asset types using the `_fetch` helper function.

Co-authored-by: AI Assistant <assistant@example.com>

Commit:
Improve readability

naychithanshwe committed:
Merge pull request #48 from TechlauncherFireApp/benmmclean/feature/asset-reference-data

Add asset type model & data layer

naychithanshwe committed:
login_view_model.dart fixed

naychithanshwe committed:
register_page.dart fixed

naychithanshwe committed:
SchedulerConstraintPage renamed

naychithanshwe committed:
SchedulerConstraintForm renamed

naychithanshwe committed:
view model renamed

naychithanshwe committed:
submit form changed

naychithanshwe committed:
app_en.arb fixed

naychithanshwe committed:
Merge remote-tracking branch 'origin/NayChiThanShwe/Feature/Reimplement-constraint-form' into NayChiThanShwe/Feature/Reimplement-constraint-form

naychithanshwe committed:
asset types added

naychithanshwe committed:
set common border as private member in BaseInputField

naychithanshwe committed:
added temporal class

naychithanshwe committed:
added scheduler infront of class names

naychithanshwe committed:
Merge branch 'main' into NayChiThanShwe/Feature/Reimplement-constraint-form

emilymclean commented on fireapp/lib/main.dart:
```diff
@@ -49,7 +53,7 @@ class _MyAppState extends State<MyApp> {
       ],
       title: 'FireApp',
-      initialRoute: (EnvironmentConfig.serviceUrl.isEmpty) ? '/server_url_page' : '/login',
+      initialRoute: '/constraint_form',
```
Please don't commit this change

emilymclean commented on fireapp/lib/presentation/constraint_form/constraint_form_view.dart:
```diff
@@ -0,0 +1,132 @@
+              icon: Icons.title,
+              validator: (v) => v!.isEmpty ? 'Title is empty!' : null,
+            ),
+            const SizedBox(height: 2.0),
```
Rather than using multiple SizedBox, you should call `spacedBy()` on the array of Column children.

emilymclean commented on fireapp/lib/presentation/constraint_form/constraint_form_view.dart:
```diff
@@ -0,0 +1,132 @@
+                        value: value,
+                        child: Text(
+                          'Asset $value',
+                          style: Theme.of(context).textTheme.labelLarge?.copyWith(
```
Do not copy the theme with hard coded colours. You should be using colours defined in the theme. Find the closest colour or add it to the theme

emilymclean commented on fireapp/lib/presentation/constraint_form/constraint_form_view.dart:
```diff
@@ -0,0 +1,132 @@
+    return Form(
+      key: viewModel.formKey,
+      child: Padding(
+        padding: const EdgeInsets.all(12.0),
```
Express spacing as an rdp. This should be 1.rdp(). An rdp is 16 units.

emilymclean commented on fireapp/lib/presentation/constraint_form/constraint_form_view.dart:
```diff
@@ -0,0 +1,132 @@
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(title: const Text('Scheduler')),
```
Localise the string

emilymclean commented on fireapp/lib/presentation/constraint_form/date_input_field.dart:
```diff
@@ -0,0 +1,31 @@
+        lastDate: DateTime(2100),
+      );
+      if (selectedDate != null) {
+        controller.text = DateFormat('yyyy-MM-dd').format(selectedDate);
```
This shouldn't be an inline date format. Make this a constant in base, call it something like "presentableDate".

emilymclean commented on fireapp/lib/presentation/constraint_form/date_input_field.dart:
```diff
@@ -0,0 +1,31 @@
+      DateTime? selectedDate = await showDatePicker(
+        context: context,
+        initialDate: DateTime.now(),
+        firstDate: DateTime(2020),
```
Put these in constants as well

emilymclean commented on fireapp/lib/presentation/constraint_form/temporal_input_field.dart:
```diff
@@ -0,0 +1,29 @@
+        // using the same text theme from login_page
+        decoration: InputDecoration(
+          labelText: AppLocalizations.of(context)?.enterDate,
+          fillColor: Colors.white,
```
You should be accessing this kind of information through Theme. If we wanted to add a dark mode, or a high contrast mode, these constants would prevent us from doing so.

emilymclean commented on fireapp/lib/presentation/constraint_form/text_input_field.dart:
```diff
@@ -0,0 +1,32 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/src/widgets/framework.dart';
+
+class SchedulerTextInputField extends BaseInputField {
```
Just name this TextInputField (unless that name is already taken, then call it FireAppTextInputField). You've written a nice, generalised class here that can be used across the whole app, don't sell it short!

emilymclean commented on fireapp/lib/presentation/constraint_form/base_input_field.dart:
```diff
@@ -0,0 +1,29 @@
+    required this.icon,
+    this.validator,
+    this.style,
+  });
```
Write a function that builds an InputDecoration object, and add abstract fields for startIcon, endIcon, hint, and label. Then have each implementing class override those fields and call the function that builds InputDecoration, rather than building it themselves.

emilymclean reviewed (changes requested):
Most of my feedback is that you need to stop using magic numbers and ad-hoc styles. Create constants, use translatable strings, and use the style framework.

Beyond that, there are some minor tweaks to the Input classes I would like, but you're pretty close.

naychithanshwe committed:
TextInputField renamed

naychithanshwe committed:
AppBar Scheduler text localized

naychithanshwe committed:
Spacing converted to rdp

naychithanshwe committed:
render flex error fixed

naychithanshwe committed:
SizedBox repalced with Spacer()

naychithanshwe committed:
rdp() fixed

naychithanshwe committed:
new color added to theme and hard coded colors replaced with colors from theme.dart

naychithanshwe committed:
presentableDate created

naychithanshwe committed:
date constants created

naychithanshwe committed:
Merge remote-tracking branch 'origin/main' into NayChiThanShwe/Feature/Reimplement-constraint-form

naychithanshwe committed:
Merge remote-tracking branch 'origin/NayChiThanShwe/Feature/Reimplement-constraint-form' into NayChiThanShwe/Feature/Reimplement-constraint-form

naychithanshwe committed:
build inputDecoration function

naychithanshwe committed:
temporal_input_field.dart and text_input_field.dart fixed

naychithanshwe committed:
unit test added for constraint_form_view_model.dart

naychithanshwe committed:
unit test fixed and mockito test generated for constraint_form_view_model_test.dart

naychithanshwe committed:
Merge remote-tracking branch 'origin/main' into NayChiThanShwe/Feature/Reimplement-constraint-form

naychithanshwe committed:
matcher added

Commit:
Fixed test dependency

Commit:
Update pubspec.yaml

emilymclean commented on fireapp/test/presentation/constraint_form/constraint_form_view_model_test.dart:
```diff
@@ -0,0 +1,110 @@
+  group('SchedulerConstraintFormViewModel Tests', () {
+
+    late SchedulerConstraintFormViewModel viewModel;
+    late MockGlobalKey<FormState> mockFormKey;
```
Don't mock these

naychithanshwe committed:
Auto stash before merge of "NayChiThanShwe/Feature/Reimplement-constraint-form" and "origin/NayChiThanShwe/Feature/Reimplement-constraint-form"

naychithanshwe committed:
removed mocked in test case

naychithanshwe committed:
Merge remote-tracking branch 'origin/main' into NayChiThanShwe/Feature/Reimplement-constraint-form

naychithanshwe committed:
duplicates removed in app_en.arb

naychithanshwe committed:
, added in main.dart

naychithanshwe committed:
fixed

Commit:
Fixed tests

Commit:
Fixed remaining issues

emilymclean committed:
Delete fireapp/flutter

Commit:
Ignore bit that can't be unit tested

emilymclean committed:
Update unit.yml
